File: ios/LiulianUI/Sources/LiulianUI/LiulianInput.swift
import SwiftUI

/// LiulianInput — per ui-spec/Input.spec.md.
public struct LiulianInput: View {
    public enum Variant {
        case text, password, search, number, textarea
    }
    public enum Size {
        case sm, md
    }

    @Binding var value: String
    let variant: Variant
    let size: Size
    let label: String?
    let placeholder: String?
    let helpText: String?
    let errorText: String?
    let disabled: Bool
    let readonly: Bool

    @FocusState private var isFocused: Bool

    public init(
        value: Binding<String>,
        variant: Variant = .text,
        size: Size = .md,
        label: String? = nil,
        placeholder: String? = nil,
        helpText: String? = nil,
        errorText: String? = nil,
        disabled: Bool = false,
        readonly: Bool = false
    ) {
        self._value = value
        self.variant = variant
        self.size = size
        self.label = label
        self.placeholder = placeholder
        self.helpText = helpText
        self.errorText = errorText
        self.disabled = disabled
        self.readonly = readonly
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: LiulianTokens.Spacing.s2) {
            if let label = label {
                LiulianText(label, variant: .bodyStrong, color: LiulianTokens.Colors.inkCharcoal)
            }
            inputField
            if let errorText = errorText {
                LiulianText(errorText, variant: .caption, color: LiulianTokens.Colors.unibeRedText)
            } else if let helpText = helpText {
                LiulianText(helpText, variant: .caption, color: LiulianTokens.Colors.inkMuted)
            }
        }
    }

    private var inputField: some View {
        let bg = (disabled || readonly) ? LiulianTokens.Colors.surfaceShade : LiulianTokens.Colors.surfacePure
        let borderColor: Color = {
            if errorText != nil { return LiulianTokens.Colors.unibeRedText }
            if isFocused { return LiulianTokens.Colors.unibeRed }
            return LiulianTokens.Colors.hairline
        }()
        let height: CGFloat = {
            if variant == .textarea { return 80 }
            return size == .sm ? LiulianTokens.Control.Height.sm : LiulianTokens.Control.Height.md
        }()

        return Group {
            if variant == .textarea {
                TextEditor(text: $value)
                    .focused($isFocused)
                    .disabled(disabled)
                    .frame(height: height)
                    .padding(.horizontal, LiulianTokens.Spacing.s3)
                    .padding(.vertical, LiulianTokens.Spacing.s2)
                    .scrollContentBackground(.hidden)
            } else if variant == .password {
                SecureField(placeholder ?? "", text: $value)
                    .focused($isFocused)
                    .disabled(disabled)
                    .frame(height: height)
                    .padding(.horizontal, LiulianTokens.Spacing.s3)
            } else {
                TextField(placeholder ?? "", text: $value)
                    .focused($isFocused)
                    .disabled(disabled || readonly)
                    .keyboardType(variant == .number ? .numberPad : .default)
                    .frame(height: height)
                    .padding(.horizontal, LiulianTokens.Spacing.s3)
            }
        }
        .font(LiulianFont.body(LiulianTokens.FontSize.md))
        .foregroundColor(LiulianTokens.Colors.inkCharcoal)
        .accentColor(LiulianTokens.Colors.unibeRed)
        .background(bg)
        .overlay(
            RoundedRectangle(cornerRadius: LiulianTokens.Radius.sm)
                .strokeBorder(borderColor, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: LiulianTokens.Radius.sm))
        .opacity(disabled ? LiulianTokens.Opacity.disabled : 1.0)
    }
}

#if DEBUG
struct LiulianInput_Previews: PreviewProvider {
    struct PreviewContainer: View {
        @State var text1 = ""
        @State var text2 = "abc"
        @State var text3 = "read-only"
        @State var text4 = "disabled"

        var body: some View {
            VStack(alignment: .leading, spacing: LiulianTokens.Spacing.s4) {
                LiulianInput(value: $text1, label: "Station", placeholder: "e.g. aare-bern")
                LiulianInput(value: $text2, label: "Validated", errorText: "Must start with a letter")
                LiulianInput(value: $text3, label: "Readonly", readonly: true)
                LiulianInput(value: $text4, label: "Disabled", disabled: true)
            }
            .padding(LiulianTokens.Spacing.s6)
            .background(LiulianTokens.Colors.canvasWarm)
        }
    }
    static var previews: some View {
        PreviewContainer()
    }
}
#endif

File: android/liulian-ui/src/main/kotlin/io/liulian/ui/LiulianInput.kt
package io.liulian.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.foundation.text.KeyboardOptions
import io.liulian.tokens.LiulianTokens

enum class InputVariant { Text, Password, Search, Number, Textarea }
enum class InputSize { Sm, Md }

/**
 * LiulianInput — Compose impl per ui-spec/Input.spec.md.
 *
 * BasicTextField wrapped in own bordered Box. No Material TextField — avoids ripple
 * + lets us control border color on focus/error precisely.
 */
@Composable
fun LiulianInput(
    value: String,
    onChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    variant: InputVariant = InputVariant.Text,
    size: InputSize = InputSize.Md,
    label: String? = null,
    placeholder: String? = null,
    helpText: String? = null,
    errorText: String? = null,
    disabled: Boolean = false,
    readonly: Boolean = false,
) {
    var focused by remember { mutableStateOf(false) }
    val hasError = errorText != null

    val borderColor = when {
        hasError -> LiulianTokens.Colors.unibeRedText
        focused -> LiulianTokens.Colors.unibeRed
        else -> LiulianTokens.Colors.hairline
    }
    val bg = if (disabled || readonly) LiulianTokens.Colors.surfaceShade else LiulianTokens.Colors.surfacePure

    val height = if (variant == InputVariant.Textarea) 80.dp else when (size) {
        InputSize.Sm -> LiulianTokens.Control.Height.sm
        InputSize.Md -> LiulianTokens.Control.Height.md
    }

    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(LiulianTokens.Spacing.s2),
    ) {
        if (label != null) {
            LiulianText(label, variant = LiulianTextVariant.BodyStrong, color = LiulianTokens.Colors.inkCharcoal)
        }
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(height)
                .clip(RoundedCornerShape(LiulianTokens.Radius.sm))
                .background(bg)
                .border(1.dp, borderColor, RoundedCornerShape(LiulianTokens.Radius.sm))
                .padding(horizontal = LiulianTokens.Spacing.s3),
            contentAlignment = if (variant == InputVariant.Textarea) Alignment.TopStart else Alignment.CenterStart,
        ) {
            val textStyle = TextStyle(
                fontFamily = LiulianFont.body(),
                fontSize = LiulianTokens.FontSize.md,
                color = LiulianTokens.Colors.inkCharcoal,
            )
            if (value.isEmpty() && placeholder != null) {
                LiulianText(placeholder, variant = LiulianTextVariant.Body, color = LiulianTokens.Colors.inkFaint)
            }
            BasicTextField(
                value = value,
                onValueChange = { if (!disabled && !readonly) onChange(it) },
                modifier = Modifier
                    .fillMaxWidth()
                    .onFocusChanged { focused = it.isFocused },
                enabled = !disabled,
                readOnly = readonly,
                singleLine = variant != InputVariant.Textarea,
                textStyle = textStyle,
                cursorBrush = androidx.compose.ui.graphics.SolidColor(LiulianTokens.Colors.unibeRed),
                visualTransformation = if (variant == InputVariant.Password) PasswordVisualTransformation() else VisualTransformation.None,
                keyboardOptions = KeyboardOptions(
                    keyboardType = when (variant) {
                        InputVariant.Number -> KeyboardType.Number
                        InputVariant.Password -> KeyboardType.Password
                        InputVariant.Search -> KeyboardType.Text
                        else -> KeyboardType.Text
                    },
                    imeAction = if (variant == InputVariant.Search) ImeAction.Search else ImeAction.Default,
                ),
            )
        }
        if (hasError) {
            LiulianText(errorText!!, variant = LiulianTextVariant.Caption, color = LiulianTokens.Colors.unibeRedText)
        } else if (helpText != null) {
            LiulianText(helpText, variant = LiulianTextVariant.Caption, color = LiulianTokens.Colors.inkMuted)
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFFBFBFA)
@Composable
private fun LiulianInputPreview() {
    Column(
        modifier = Modifier.padding(LiulianTokens.Spacing.s6),
        verticalArrangement = Arrangement.spacedBy(LiulianTokens.Spacing.s4),
    ) {
        LiulianInput(value = "", onChange = {}, label = "Station name", placeholder = "e.g. aare-bern")
        LiulianInput(value = "secret", onChange = {}, variant = InputVariant.Password, label = "Token", helpText = "Treat as a password")
        LiulianInput(value = "abc", onChange = {}, label = "Validated", errorText = "Must start with a letter")
        LiulianInput(value = "read-only", onChange = {}, label = "Readonly", readonly = true)
        LiulianInput(value = "disabled", onChange = {}, label = "Disabled", disabled = true)
        LiulianInput(value = "", onChange = {}, variant = InputVariant.Textarea, label = "Notes", placeholder = "Longer text…")
    }
}
